crypto/context.c

```diff
@@ -160,22 +181,32 @@ void OPENSSL_CTX_free(OPENSSL_CTX *ctx)
     OPENSSL_free(ctx);
 }
 
+OPENSSL_CTX *OPENSSL_CTX_set0_default(OPENSSL_CTX *libctx)
+{
+#ifndef FIPS_MODULE
+    OPENSSL_CTX *current_defctx;
+
+    if ((current_defctx = get_default_context()) != NULL
+        && set_default_context(libctx))
+        return current_defctx;
+#endif
+
+    return NULL;
+}
+
 OPENSSL_CTX *openssl_ctx_get_concrete(OPENSSL_CTX *ctx)
 {
 #ifndef FIPS_MODULE
-    if (ctx == NULL) {
-        if (!RUN_ONCE(&default_context_init, do_default_context_init))
-            return 0;
-        return default_context;
-    }
+    if (ctx == NULL)
+        return get_default_context();
 #endif
     return ctx;
 }
 
 int openssl_ctx_is_default(OPENSSL_CTX *ctx)
```

Comment:
Given that this function is only used in two places (both in `evp_fetch()`) and both are raising an error indicating default vs non-default library context, perhaps it might indicate that a default library context was used but that it wasn't the default default one:

``` C
const char *openssl_ctx_is_default(OPENSSL_CTX *ctx)
{
    OPENSSL_CTX *dflt = get_default_context();

    if (ctx == NULL)
        ctx = dflt;
    if (ctx == &default_context_int)
        return "Default library context";
    if (ctx == dflt)
        return "Non-default default library context";
    return "Non-default library context";
}
```

Reply:
Er, so you want the caller to compare magic text to find out?

Reply:
Point taken, though, I solved that with one more function

Reply:
Done

Reply:
There is no need to compare text, the output is only ever being printed into an error message.